### PR TITLE
[WIP] MultiheadAttention building blocks

### LINKED ISSUE #34573: Restructure `multi_head_attention_forward`
## 🚀 Feature

Restructure the function `multi_head_attention_forward`  in [nn.functional](https://github.com/pytorch/pytorch/blob/23b2fba79a6d2baadbb528b58ce6adb0ea929976/torch/nn/functional.py#L3573) into several functions to improve the ability to experiment. In particular, decompose the function so that the following are available:
* The input embedding functions.
* The computation of attention weights.
* The output embedding function.

This will allow users to try different embeddings or attention mechanisms without having to recode the rest.

## Motivation

Addresses the issue of decomposing the function as mentioned in #32590. It also moves forward on including more support for attention mechanisms.

## Pitch

Currently, the `mutli_head_attention_forward` function encapsulates the projection of the `query`, `key`, and `value`, computing attention for these projections, and computing the output projection after applying attention. Furthermore, the input embedding utilizes several code paths that are different embeddings. By decomposing the function into several parts, we can make it more readable and open to experimentation.

The following breakdown is based on the above:

* Functions for computing the input embeddings `q`, `k`, and `v`. There are currently four code paths used for doing this, and each should be an individual function so that it's clearer what is the intended outcome. (See L3669-L3748.)
* A function for computing the attention weight. Some models rely on computing attention in different ways, and separating this out would allow us to use those more freely. (See L3750-L3824).
* A function for computing the output projection from the attention weight and `value` projection. (See L3826-L3836.)

### PR DESCRIPTION
**Pull request by zhangguanheng66**
Create a PR to add building blocks for MHA. Feel free to comment.
@cpuhrsch @fmassa @vincentqb 

Example to build MHA:
```
import torch
from torch.nn.modules import MultiheadAttentionInProjection, ScaledDotProduct, MultiheadAttentionOutProjection

embed, nhead = 16, 4
q_in_project = MultiheadAttentionInProjection(embed, nhead)
k_in_project = MultiheadAttentionInProjection(embed, nhead)
v_in_project = MultiheadAttentionInProjection(embed, nhead)
attention_layer = ScaledDotProduct()
out_project = MultiheadAttentionOutProjection(embed, nhead)

tgt, src, batch = 5, 10, 8
query = torch.rand((tgt, batch, embed))
q = q_in_project(query)

key = torch.rand((src, batch, embed))
k = k_in_project(key)

value = torch.rand((src, batch, embed))
v = v_in_project(value)

attn_output, attn_output_weight = attention_layer(q, k, v)
output = out_project(attn_output)
```
fix https://github.com/pytorch/pytorch/issues/34573

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
Fix https://github.com/pytorch/pytorch/issues/34537

**Review by fmassa — COMMENTED** [review]
Can you clarify in a comment what are the input-output shapes for each one of the 3 functions? I have the impression that depending on the function we will have output shape to be batch-first or not, which IMO is confusing

**Comment by zhangguanheng66** [comment]
> Can you clarify in a comment what are the input-output shapes for each one of the 3 functions? I have the impression that depending on the function we will have output shape to be batch-first or not, which IMO is confusing

Yes. That's true. We should unify those shapes across all the three building blocks. BTW, we move to #34573 for the work.

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -375,6 +375,81 @@ def forward(self, tgt, memory, tgt_mask=None, memory_mask=None,
...
+        self.dropout = dropout
+
+    def forward(self, query, key, value, attn_mask=None):
+        r"""Scaled Dot Product
+
+        Args:
+            query, key, value: map a query and a set of key-value pairs to an output.
+
+        Shape:
+            - query: :math:`(N, L, E)`
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: Mentioned by @fmassa , we should unify the input shape of those three building blocks by sequence-dimension first.

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -375,6 +375,81 @@ def forward(self, tgt, memory, tgt_mask=None, memory_mask=None,
...
+        q = torch.reshape(q, (seq, bsz * self.num_heads, head_dim))
+        return q.transpose(0, 1)
+
+
+class ScaledDotProduct(Module):
+    def __init__(self, dropout=0.0):
+        super(ScaledDotProduct, self).__init__()
+        self.dropout = dropout
+
+    def forward(self, query, key, value, attn_mask=None):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: I think this already supports broadcasting across query, key and value. This is worth verifying and documenting in my opinion.

**Pull request closed without merging**